### PR TITLE
nvidia-x11.vulkan_beta: init at 450.56.11

### PR DESCRIPTION
**Pull request by expipiplus1**
- This is fetched from a different URL, so allow passing that explicitly.

- There also isn't an nvidia-persistenced or nvidia-settings release for
  this version, so use 450.57 instead. Also implement passing
  persistenced and settings version explicitly.

See also https://github.com/NixOS/nixpkgs/pull/96057

CC @eadwu

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by eadwu — CHANGES_REQUESTED** [review]
Otherwise everything looks fine.

Didn't test it out myself but builds fine.

File: pkgs/os-specific/linux/nvidia-x11/generic.nix [hunk 1]
```diff
@@ -44,7 +47,12 @@ let
...
     src =
-      if stdenv.hostPlatform.system == "x86_64-linux" then
+      if url != null then
+        fetchurl {
+          inherit url;
+          sha256 = if stdenv.hostPlatform.is64bit then sha256_64bit else sha256_32bit;
+        }
+      else if stdenv.hostPlatform.system == "x86_64-linux" then
         fetchurl {
           url = "https://download.nvidia.com/XFree86/Linux-x86_64/${version}/NVIDIA-Linux-x86_64-${version}${pkgSuffix}.run";
```
[the remark this answers is not in the document]
Comment (expipiplus1) [reply pkgs/os-specific/linux/nvidia-x11/generic.nix]: done, thanks!

File: pkgs/os-specific/linux/nvidia-x11/default.nix [hunk 2]
```diff
@@ -32,6 +32,19 @@ rec {
...
+    then generic {
+      version = "450.56.11";
+      persistencedVersion = "450.57";
+      settingsVersion = "450.57";
+      sha256_64bit = "1k64h8sp4rf6kc7liypznjgkmxi67njy1s8xy2r341fhl62pl010";
+      settingsSha256 = "1clbj9a3kv3j8jg35c197gd7b3f9f9f4h9ll5hlax95hdg12lgan";
+      persistencedSha256 = "17747z1fsbiznfsmahxmz8kmhwwcjanpfih60v5mwzk63gy4i3d5";
+      url = https://developer.nvidia.com/vulkan-beta-4505611-linux;
+    }
+    else throw "Unsupported platform for nvidia-x11.vulkan_beta: ${stdenv.hostPlatform.system}";
```
Comment (veprbl) [inline pkgs/os-specific/linux/nvidia-x11/default.nix:46]: This breaks eval for manual. Throwing early on eval is not introspection-friendly in general. Setting `meta.platforms` and `meta.broken` allow for less restrictive use (still throws, but only on eval of the full derivation).

File: pkgs/os-specific/linux/nvidia-x11/default.nix [hunk 3]
```diff
@@ -32,6 +32,19 @@ rec {
...
+  # Beta only exists for x86_64
+  vulkan_beta = if stdenv.hostPlatform.system == "x86_64-linux"
+    then generic {
+      version = "450.56.11";
+      persistencedVersion = "450.57";
+      settingsVersion = "450.57";
+      sha256_64bit = "1k64h8sp4rf6kc7liypznjgkmxi67njy1s8xy2r341fhl62pl010";
+      settingsSha256 = "1clbj9a3kv3j8jg35c197gd7b3f9f9f4h9ll5hlax95hdg12lgan";
+      persistencedSha256 = "17747z1fsbiznfsmahxmz8kmhwwcjanpfih60v5mwzk63gy4i3d5";
+      url = https://developer.nvidia.com/vulkan-beta-4505611-linux;
```
Comment (veprbl) [inline pkgs/os-specific/linux/nvidia-x11/default.nix:44]:
```suggestion
      url = "https://developer.nvidia.com/vulkan-beta-4505611-linux";
```

File: pkgs/os-specific/linux/nvidia-x11/default.nix (hunk 2 above)
Comment (expipiplus1) [reply pkgs/os-specific/linux/nvidia-x11/default.nix:46]: Corrected (by allowing on i686, it seems as though this release does include the 32 bit libraries, I'm unable to test this though).

File: pkgs/os-specific/linux/nvidia-x11/default.nix (hunk 3 above)
Comment (expipiplus1) [reply pkgs/os-specific/linux/nvidia-x11/default.nix:44]: Thanks, corrected

File: pkgs/os-specific/linux/nvidia-x11/generic.nix [hunk 4]
```diff
@@ -1,8 +1,11 @@
 { version
+, url ? null
 , sha256_32bit ? null
 , sha256_64bit
 , settingsSha256
+, settingsVersion ? null
```
Comment (baracoder) [inline pkgs/os-specific/linux/nvidia-x11/generic.nix:6]:
```suggestion
, settingsVersion ? version
```
would remove the need for `...}@args` and `persistencedVersion = args.persistencedVersion or version;`

Comment (expipiplus1) [reply pkgs/os-specific/linux/nvidia-x11/generic.nix]: good idea! done (args is still required for `args.url or ...`)

Comment (veprbl) [reply pkgs/os-specific/linux/nvidia-x11/generic.nix]: Actually, you could compare `url` with `null`. IIRC the use of `{ }@args` with `callPackage` is discouraged, I don’t remember the reason of the top of my head.

Comment (veprbl) [reply pkgs/os-specific/linux/nvidia-x11/generic.nix]: On a second thought, `@args` should be fine as long as we don’t have an ellipsis `, ... }@args`

**Comment by expipiplus1** [comment]
I think this is ready to be merged now, I've been using it without issue for several weeks now.

**Review by baracoder — APPROVED** [review]
I don't think that this changes would break any other version.

**Comment by expipiplus1** [comment]
Thanks @veprbl

**Pull request merged**